// react-instantsearch@7.22.1/dist/es/ui/RangeInput.js
var _excluded = ["classNames", "range", "start", "step", "disabled", "onSubmit", "translations"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
import { cx } from 'instantsearch-ui-components';
import React, { useState } from 'react';
// if the default value is undefined, React considers the component uncontrolled initially, which we don't want 0 or NaN as the default value
var unsetNumberInputValue = '';

// Strips leading `0` from a positive number value
function stripLeadingZeroFromInput(value) {
  return value.replace(/^(0+)\d/, function (part) {
    return Number(part).toString();
  });
}
export function RangeInput(_ref) {
  var _values$min, _values$max;
  var _ref$classNames = _ref.classNames,
    classNames = _ref$classNames === void 0 ? {} : _ref$classNames,
    _ref$range = _ref.range,
    min = _ref$range.min,
    max = _ref$range.max,
    _ref$start = _slicedToArray(_ref.start, 2),
    minValue = _ref$start[0],
    maxValue = _ref$start[1],
    _ref$step = _ref.step,
    step = _ref$step === void 0 ? 1 : _ref$step,
    disabled = _ref.disabled,
    _onSubmit = _ref.onSubmit,
    translations = _ref.translations,
    props = _objectWithoutProperties(_ref, _excluded);
  var values = {
    min: minValue !== -Infinity && minValue !== min ? minValue : unsetNumberInputValue,
    max: maxValue !== Infinity && maxValue !== max ? maxValue : unsetNumberInputValue
  };
  var _useState = useState(values),
    _useState2 = _slicedToArray(_useState, 2),
    prevValues = _useState2[0],
    setPrevValues = _useState2[1];
  var _useState3 = useState({
      from: (_values$min = values.min) === null || _values$min === void 0 ? void 0 : _values$min.toString(),
      to: (_values$max = values.max) === null || _values$max === void 0 ? void 0 : _values$max.toString()
    }),
    _useState4 = _slicedToArray(_useState3, 2),
    _useState4$ = _useState4[0],
    from = _useState4$.from,
    to = _useState4$.to,
    setRange = _useState4[1];
  if (values.min !== prevValues.min || values.max !== prevValues.max) {
    var _values$min2, _values$max2;
    setRange({
      from: (_values$min2 = values.min) === null || _values$min2 === void 0 ? void 0 : _values$min2.toString(),
      to: (_values$max2 = values.max) === null || _values$max2 === void 0 ? void 0 : _values$max2.toString()
    });
    setPrevValues(values);
  }
  return /*#__PURE__*/React.createElement("div", _extends({}, props, {
    className: cx(cx('ais-RangeInput', classNames.root), disabled && cx('ais-RangeInput--noRefinement', classNames.noRefinementRoot), props.className)
  }), /*#__PURE__*/React.createElement("form", {
    className: cx('ais-RangeInput-form', classNames.form),
    onSubmit: function onSubmit(event) {
      event.preventDefault();
      _onSubmit([from ? Number(from) : undefined, to ? Number(to) : undefined]);
    }
  }, /*#__PURE__*/React.createElement("label", {
    className: cx('ais-RangeInput-label', classNames.label)
  }, /*#__PURE__*/React.createElement("input", {
    className: cx('ais-RangeInput-input', classNames.input, 'ais-RangeInput-input--min', classNames.inputMin),
    type: "number",
    min: min,
    max: max,
    value: stripLeadingZeroFromInput(from || unsetNumberInputValue),
    step: step,
    placeholder: min === null || min === void 0 ? void 0 : min.toString(),
    disabled: disabled,
    onInput: function onInput(_ref2) {
      var currentTarget = _ref2.currentTarget;
      var value = currentTarget.value;
      setRange({
        from: value || unsetNumberInputValue,
        to: to
      });
    }
  })), /*#__PURE__*/React.createElement("span", {
    className: cx('ais-RangeInput-separator', classNames.separator)
  }, translations.separatorElementText), /*#__PURE__*/React.createElement("label", {
    className: cx('ais-RangeInput-label', classNames.label)
  }, /*#__PURE__*/React.createElement("input", {
    className: cx('ais-RangeInput-input', classNames.input, 'ais-RangeInput-input--max', classNames.inputMax),
    type: "number",
    min: min,
    max: max,
    value: stripLeadingZeroFromInput(to || unsetNumberInputValue),
    step: step,
    placeholder: max === null || max === void 0 ? void 0 : max.toString(),
    disabled: disabled,
    onInput: function onInput(_ref3) {
      var currentTarget = _ref3.currentTarget;
      var value = currentTarget.value;
      setRange({
        from: from,
        to: value || unsetNumberInputValue
      });
    }
  })), /*#__PURE__*/React.createElement("button", {
    className: cx('ais-RangeInput-submit', classNames.submit),
    type: "submit"
  }, translations.submitButtonText)));
}
