// react-instantsearch@7.22.1/dist/es/widgets/HitsPerPage.js
var _excluded = ["items", "transformItems"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
import React from 'react';
import { useHitsPerPage } from 'react-instantsearch-core';
import { HitsPerPage as HitsPerPageUiComponent } from "../ui/HitsPerPage.js";
export function HitsPerPage(_ref) {
  var userItems = _ref.items,
    transformItems = _ref.transformItems,
    props = _objectWithoutProperties(_ref, _excluded);
  var _useHitsPerPage = useHitsPerPage({
      items: userItems,
      transformItems: transformItems
    }, {
      $$widgetType: 'ais.hitsPerPage'
    }),
    items = _useHitsPerPage.items,
    refine = _useHitsPerPage.refine;
  var _ref2 = items.find(function (_ref3) {
      var isRefined = _ref3.isRefined;
      return isRefined;
    }) || {},
    currentValue = _ref2.value;
  var uiProps = {
    items: items,
    currentValue: currentValue,
    onChange: function onChange(value) {
      return refine(value);
    }
  };
  return /*#__PURE__*/React.createElement(HitsPerPageUiComponent, _extends({}, props, uiProps));
}
